File: rust/src/display/canvas.rs
use crate::display::color::Color;
use crate::display::resolution::Resolution;

pub struct Canvas {
    pub width: usize,
    pub height: usize,
    pub pixels: Vec<Color>,
}

impl Canvas {
    pub fn new(resolution: Resolution) -> Canvas {
        let width = resolution.hsize();
        let height = resolution.vsize();
        Canvas {
            width,
            height,
            pixels: vec![Color::BLACK; width * height],
        }
    }

    pub fn rows(&self) -> Vec<Vec<Color>> {
        (0..self.height)
            .map(|y| (0..self.width).map(|x| self.pixel_at(x, y)).collect())
            .collect()
    }

    pub fn write_pixel(&mut self, x: usize, y: usize, color: Color) {
        let index = self.index_for(x, y);
        self.pixels[index] = color;
    }

    pub fn pixel_at(&self, x: usize, y: usize) -> Color {
        self.pixels[self.index_for(x, y)]
    }

    fn index_for(&self, x: usize, y: usize) -> usize {
        if x > self.width || y > self.height {
            panic!(
                "coordinate ({}, {}) is lies outside canvas size: {}x{}",
                x, y, self.width, self.height
            );
        }
        y * self.width + x
    }
}

File: kotlin/src/main/kotlin/display/Canvas.kt
package display

import java.lang.RuntimeException

data class Canvas(val resolution: Resolution) {

    private val pixels = MutableList(resolution.hsize * resolution.vsize) { Color.BLACK }

    fun pixelAt(x: Int, y: Int) = pixels[indexFor(x, y)]

    fun writePixel(x: Int, y: Int, color: Color) {
        pixels[indexFor(x, y)] = color
    }

    /**
     * Returns a list of rows, each of which is a list of Colors.
     */
    fun rows() = (0 until resolution.vsize).map { y -> (0 until resolution.hsize).map { x -> pixelAt(x, y) } }

    /**
     * Determine the list index for the specified coordinates.
     */
    private fun indexFor(x: Int, y: Int): Int {
        if (x < 0 || x > resolution.hsize || y < 0 || y > resolution.vsize) {
            throw CanvasException("coordinate ($x, $y) is lies outside canvas size: ${resolution.hsize}x${resolution.vsize}")
        }

        return y * resolution.hsize + x
    }

    class CanvasException(message: String) : RuntimeException(message)
}
